### PR TITLE
bpo-46475: Add typing.Never and typing.assert_never

### PR DESCRIPTION
**Pull request by JelleZijlstra**
Another pending proposal from typing-sig.

- Add a `Never` primitive that is semantically identical to `NoReturn`, but documented explicitly as a bottom type. I chose to make it a separate object instead of a pure alias for `NoReturn` so that dynamic checkers can distinguish between the two and so that we can document them separately.
- Add an `assert_never()` function to statically assert that code is unreachable. Based on @davidfstr's feedback, the error raised by the function does not include the argument.

https://bugs.python.org/issue46475

### CODE REVIEW

File: Doc/library/typing.rst
Comment (sobolevn) [inline Doc/library/typing.rst:605]: Should we deprecate it?

File: Doc/library/typing.rst
Comment (sobolevn) [inline Doc/library/typing.rst]: I am not sure about the `match` semantics: how `arg` will be narrowed, but maybe we can recommend `case _ as unreachable: assert_never(unreachable)`?

File: Lib/test/test_typing.py
Comment (sobolevn) [inline Lib/test/test_typing.py]: I think we can create something like `_BottomTypeTests` base class and then subclass it with `NoReturnTests` and `NeverTests`.

File: Lib/typing.py
Comment (sobolevn) [inline Lib/typing.py:458]: Maybe it is a good idea to somehow alias `NoReturn`? 🤔

**Review by sobolevn — COMMENTED** [review]
Thank you! Awesome features!

File: Doc/library/typing.rst
Comment (JelleZijlstra) [reply Doc/library/typing.rst:605]: @erictraut's suggestion was to allow type checkers to limit NoReturn to what it's meant for (functions that don't return) in the future.

File: Doc/library/typing.rst
Comment (JelleZijlstra) [reply Doc/library/typing.rst]: Not sure why that would be better.

File: Lib/typing.py
Comment (JelleZijlstra) [reply Lib/typing.py:458]: I'd rather not alias. I covered this in my PR description.

File: Lib/test/test_typing.py
Comment (JelleZijlstra) [reply Lib/test/test_typing.py]: Good idea done. I didn't underscore it by analogy with `NewTypeTests`.

File: Doc/library/typing.rst
Comment (Fidget-Spinner) [inline Doc/library/typing.rst]: I just noticed the Wikipedia page says `typing.NoReturn` is Python's bottom type. We live in a strange world.

File: Doc/library/typing.rst
Comment (davidfstr) [reply Doc/library/typing.rst:605]: > limit NoReturn to what it's meant for (functions that don't return) in the future.

Introducing such a type checker limitation (which I am ±0 on) would make it impossible to spell "bottom type" in code that must support Python <3.11, unless we **also add typing_extensions.Never**, which I do recommend.

File: Doc/library/typing.rst
Comment (davidfstr) [reply Doc/library/typing.rst]: For a more complex `match` expression, I _do_ think the `case _ as unreachable: assert_never(unreachable)` phrasing would work better. Consider:

```
match type(foo):
    case type(int)():
        ...
    case type(str)():
        ...
    case _ as unreachable:
        assert_never(unreachable)
```

File: Lib/typing.py
Comment (davidfstr) [inline Lib/typing.py]: Maybe use a message like "Expected code to be unreachable" ? Clearly it's not unreachable.

(Aside: This kind of runtime error I expect to occur if someone writes code that is NOT actually unreachable and hasn't yet run a typechecker on it, which would complain.)

File: Doc/library/typing.rst
Comment (davidfstr) [reply Doc/library/typing.rst]: Seems like we'll be able to update that page soon :)

File: Doc/library/typing.rst
Comment (JelleZijlstra) [reply Doc/library/typing.rst:605]: I'm also not sure we should do that, but it's nice to have the option.

I will add `typing_extensions.Never` once this PR is merged. (With some luck it will be a very feature-packed typing-extensions release: PEP 646, `@dataclass_transform`, and all the little functions I'm adding.)

File: Lib/typing.py
Comment (JelleZijlstra) [reply Lib/typing.py]: Yes, that's better, thanks. I'll update the PR tomorrow.

File: Doc/library/typing.rst
Comment (JelleZijlstra) [reply Doc/library/typing.rst]: Done

File: Lib/test/test_typing.py
Comment (gvanrossum) [inline Lib/test/test_typing.py]: I would rather name this to emphasize it's a mixin class; I was for a while mystified by the presence of an uninitialized bottom_type variable.

**Review by gvanrossum — COMMENTED** [review]
I think we should iterate this until @Fidget-Spinner approves it and then he can merge it. I think the idea is good, and IIRC typing-sig is in agreement. (Though are there folks at Pyre or Google who don't read it but want issues instead? Sigh. :-( )

File: Lib/test/test_typing.py
Comment (JelleZijlstra) [reply Lib/test/test_typing.py]: Good call, I renamed it.

**Comment by JelleZijlstra** [comment]
Thanks @gvanrossum! I think I addressed all previous comments, so from my perspective this is ready.

https://mail.python.org/archives/list/typing-sig@python.org/thread/CGIYKGMF6XLDX3F2JNRCCE7HC6K2XLZ2/ is the typing-sig thread. Nobody opposed it.

**Comment by gvanrossum** [comment]
Should I wait for the tests or are they hanging?

File: Doc/library/typing.rst
Comment (gvanrossum) [inline Doc/library/typing.rst]: It feels wrong to only show examples of errors. Please give a positive example too (e.g. a function that raises immediately). Maybe only give a positive example?

**Comment by JelleZijlstra** [comment]
They just started, I had to fix a merge conflict. It's going to conflict with the Self change too, which is why I pinged you on that one first. I think this is the easiest way forward:
- We wait for the tests on Self to pass and merge it
- I fix the conflict here and wait for tests to pass again
- Then we merge this one

File: Doc/library/typing.rst
Comment (gvanrossum) [inline Doc/library/typing.rst]: Is that relevant here? What does "statically assert" even mean? (We don't use this phrase elsewhere in this file.) I know that the point is that its definition is
```py
def assert_never() -> Never:
    ...
```
but if the reader doesn't already know that (and why would they if they are reading this?) they wouldn't figure it out from the phrasing here.

File: Doc/library/typing.rst
Comment (gvanrossum) [reply Doc/library/typing.rst:605]: > I'm also not sure we should do that, but it's nice to have the option.

Do what? Deprecate using `NoReturn` outside return types?

Honestly I feel there's very little reason to deprecate that given that it apparently used to work. Too much churn isn't good.

File: Doc/library/typing.rst
Comment (gvanrossum) [inline Doc/library/typing.rst]: Why not just say that (a) this usage is equivalent to a bottom type, and (b) it is preferred to use `Never` as a bottom type (but only in 3.11+)?

File: Lib/typing.py
Comment (gvanrossum) [inline Lib/typing.py]: Has it already been litigated whether `AssertionError` is a better exception to raise here?

File: Doc/library/typing.rst
Comment (gvanrossum) [inline Doc/library/typing.rst]: Oh, I'm guessing that "statically assert X" means "ask the type checker to verify that X is provable". If this is standard terminology, maybe this phrase should be explained somewhere?

File: Lib/typing.py
Comment (gvanrossum) [inline Lib/typing.py]: Maybe drop "general", since the original intention was more narrow?

File: Lib/typing.py
Comment (gvanrossum) [reply Lib/typing.py:458]: Can you please add a comment explaining the motivation? Otherwise next year someone in pursuit of a quick reduction in lines of code will merge the two. :-)

File: Lib/typing.py
Comment (gvanrossum) [inline Lib/typing.py]: I'd drop "Notation for". This *is* the bottom type, right? In the end, *everything* is a notation for some concept, but we don't normally use that phrase.

File: Lib/typing.py
Comment (gvanrossum) [inline Lib/typing.py]: Does the docstring need to say this? Does assert_never() need this much advertising? :-)

File: Lib/typing.py
Comment (gvanrossum) [inline Lib/typing.py:2101]: This example is much better than the examples in the main doc. :-) And so is the sentence following the example.

**Review by gvanrossum — COMMENTED** [review]
Sorry, I've got more...

File: Doc/library/typing.rst
Comment (JelleZijlstra) [reply Doc/library/typing.rst:605]: Yes, that would be the suggestion, and agree that it's not a very valuable change. Sounds like there's no action item here, so I'll resolve this thread.

File: Lib/typing.py
Comment (JelleZijlstra) [reply Lib/typing.py:458]: Done

File: Doc/library/typing.rst
Comment (JelleZijlstra) [reply Doc/library/typing.rst]: Changed this to also give the `int_or_str` example from the `assert_never` docs.

File: Doc/library/typing.rst
Comment (JelleZijlstra) [reply Doc/library/typing.rst]: Removed the mention of `assert_never`.

File: Doc/library/typing.rst
Comment (JelleZijlstra) [reply Doc/library/typing.rst]: Reworded.

File: Lib/typing.py
Comment (JelleZijlstra) [reply Lib/typing.py]: Hasn't been discussed yet. Personally I don't like raising AssertionError manually; I feel it should only be used from the actual `assert` statement.

File: Doc/library/typing.rst
Comment (JelleZijlstra) [reply Doc/library/typing.rst]: Well, C++ has a `static_assert()` function. But it's better to avoid the term if it's confusing, so I'll reword to "assert to the type checker".

File: Lib/typing.py
Comment (JelleZijlstra) [reply Lib/typing.py:2101]: That sentence is also in the docs for assert_never, but not Never. I'm pushing some changes to improve the Never docs according to your feedback.

Honestly I'm a bit confused about the relationship between the docstring and the RST docs. I end up copying from one to the other.

File: Lib/typing.py
Comment (AlexWaygood) [reply Lib/typing.py]: I quite like the idea of using `AssertionError` here, actually. `assert`s and `AssertionError`s are for places in code that you hope will never be reached, if everything's working the way you expected it to. If you hit an `AssertionError`, you know there's been a logical error in the code that was written. And that feels very fitting for this situation, whereas a `RuntimeError` could be caused by all sorts of things.

You could always use `assert False` if you don't like the way `raise AssertionError` _looks_ ;)

Comment (erictraut) [reply Lib/typing.py]: The `assert` statement becomes a no-op if `__debug__` is False. Would it be better if `assert_never` does the same?

**Comment by JelleZijlstra** [comment]
Thanks for your feedback! I pushed a round of docs changes.

File: Doc/library/typing.rst
Comment (AlexWaygood) [inline Doc/library/typing.rst]: Are we deprecating `NoReturn`? If not, in which cases are we saying `NoReturn` should still be used — surely this is a textbook example of a situation where the `NoReturn` name is more descriptive of the semantics?

File: Lib/typing.py
Comment (AlexWaygood) [inline Lib/typing.py]: There's repetition here on lines 458-461, and using Rest syntax in a docstring is a little weird

Comment (JelleZijlstra) [reply Lib/typing.py]: Good catch, will remove

File: Doc/library/typing.rst
Comment (JelleZijlstra) [reply Doc/library/typing.rst]: Good point, I'll remove this second example.

Comment (AlexWaygood) [reply Doc/library/typing.rst]: Looks like the same example's on line 478 of `typing.py` as well (in the docstring of `Never`)

File: Doc/library/typing.rst
Comment (AlexWaygood) [inline Doc/library/typing.rst:583]: I would scrap the "or a function that never returns" part here, if we want people to carry on using the `NoReturn` name for that purpose. Let's be really clear about the intended difference between the two.

Comment (gvanrossum) [reply Doc/library/typing.rst:583]: No, we should not encourage users to use `NoReturn`. We should just not kill it. I think `-> Never` is just as clear as `-> NoReturn` -- possibly clearer.

Comment (AlexWaygood) [reply Doc/library/typing.rst:583]: Perhaps it's the `versionadded` section below that's making me uncomfortable — the "on older Python versions" wording makes it sound (to me) like it's saying "using `NoReturn` to do any of the above things is now discouraged ". Which sounds a lot like deprecation — but we're not deprecating.

I agree that `-> Never` is just as clear.

Comment (gvanrossum) [reply Doc/library/typing.rst:583]: No, discouraging is not the same as deprecating. (E.g. PEP 8 discourages lots of things that are not and will never be deprecated.) I think we *should* nudge people in the direction of `Never` (TOOWTDI and all that), but we needn't deprecate `NoReturn` until 3.10 is a distant memory.

Comment (AlexWaygood) [reply Doc/library/typing.rst:583]: Okay.

File: Doc/library/typing.rst
Comment (JelleZijlstra) [reply Doc/library/typing.rst]: Thanks, fixed that one too

File: Lib/typing.py
Comment (JelleZijlstra) [reply Lib/typing.py]: I switched it to AssertionError.

I don't think making the behavior rely on `__debug__` makes sense here: that's a performance optimization to strip out assertions so you don't pay the cost of checking them, but if you hit an assert_never(), there's nothing else to do.

Comment (gvanrossum) [reply Lib/typing.py]: Correct, when raising `AssertionError` manually you generally shouldn't check `__debug__`. (The latter exists so you can write code more elaborate than a bunch of `assert` statements without having to turn it into a function.)

**Review by AlexWaygood — APPROVED** [review]
LGTM — thanks!

File: Lib/typing.py
Comment (AlexWaygood) [inline Lib/typing.py:2106]: Have you given any thought to potentially making the runtime error message user-customisable? You could have something like this:

```python
def assert_never(arg: Never, /, *, msg=None) -> Never:
    if msg is None:
        msg = "Expected code to be unreachable"
    raise AssertionError(msg)
```

Comment (JelleZijlstra) [reply Lib/typing.py:2106]: I don't want to complicate this thing prematurely. I'd rather we land this change so we can move forward with putting it in typing-extensions and type checkers.

Comment (AlexWaygood) [reply Lib/typing.py:2106]: Sure 👍

File: Lib/typing.py
Comment (gvanrossum) [reply Lib/typing.py:2101]: Usually the docstring is shorter, since it is often displayed by tools that have limited output space (it's really annoying to get a scroll bar in a tooltip in VS Code :-).

**Review by gvanrossum — APPROVED** [review]
Great! Thanks for being patient with me.

**Comment by JelleZijlstra** [comment]
Thank you for the feedback!

File: Doc/library/typing.rst
Comment (henryiii) [reply Doc/library/typing.rst]: Is there a reason it's:

```python
    case _ as unreachable:
        assert_never(unreachable)
```

Instead of the slightly shorter:

```python
    case unreachable:
        assert_never(unreachable)
```

?

Comment (JelleZijlstra) [reply Doc/library/typing.rst]: No strong reason, but I feel like this makes it slightly more clear that it's a fallback case.

Comment (davidfstr) [reply Doc/library/typing.rst]: Agreed that the `case _ as unreachable` phrasing is more clearly intended as a catch-all than just `case unreachable`.

Comment (gvanrossum) [reply Doc/library/typing.rst]: I'm not sure I agree. It's important to teach/remind people that an unqualified capture like `case unreachable:` must (if it occurs) always be the last case. (In fact the parser warns if anything follows.)

**Pull request merged**
